waspc/src/Wasp/Generator/Job/IO/PrefixedWriter.hs

```diff
@@ -0,0 +1,125 @@
+
+module Wasp.Generator.Job.IO.PrefixedWriter
+  ( printPrefixedJobMessage,
+    runPrefixedWriter,
+    PrefixedWriter,
+  )
+where
+
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.State (gets, modify)
+import Control.Monad.State.Strict (MonadState, StateT, runStateT)
+import qualified Data.Set as S
+import qualified Data.Text as T
+import qualified Data.Text.IO as T.IO
+import System.IO (hFlush, stderr)
+import qualified Wasp.Generator.Job as J
+import Wasp.Generator.Job.Common (getJobMessageContent, getJobMessageOutHandle)
+import qualified Wasp.Util.Terminal as Term
+
+-- |
+-- Imagine you have a job sending following two messages:
+--  1. "First"
+--  2. " line\n"
+--  3. "Second line"
+--
+-- What we want is to prefix new lines with a corresponding job prefix before we print them,
+-- e.g. "Server: ". So we want to output this as:
+--   Server: First line
+--   Server: Second line
+--
+-- This is what this function does, it properly prefixes the given message and then prints it.
+--
+-- * Implementation details:
+--
+-- Simplest (and naive) way to go about this is to add prefix after any newline.
+-- However, what can happen is that third message from above comes quite later than the second
+-- message, and in the meantime some other output (e.g. from another job) is printed.
+-- In such case, we get following:
+--   Server: First line
+--   Server:
+--   Some other output!
+--   Second line
+--
+-- This is not an easy problem to solve, as we can't know what kind of message is coming next,
+-- and there are always situations where some other output might interrupt us.
+-- But, what we can at least do is avoid having this situation described above, where newline is
+-- "kidnapped", and we do that by postponing the newline (make it pending) for later,
+-- until the next message from the same job arrives.
+--
+-- Specifically, what we do is postpone printing of a newline if it is the last character in a message.
+-- We make it pending instead, and once the new message comes from the same job, we apply it at the
+-- start of that message.
+--
+-- This way we get proper output in the situation as descried above:
+--   Server: First line
+--   Some other output!
+--   Server: Second line
+printPrefixedJobMessage :: J.JobMessage -> PrefixedWriter ()
+printPrefixedJobMessage jobMessage = do
+  jobsWithPendingNewline <- gets _jobsWithPendingNewline
+  let (jobsWithPendingNewline', content') =
+        applyPendingNewline jobsWithPendingNewline (jobType, content)
+  modify $ \s -> s {_jobsWithPendingNewline = jobsWithPendingNewline'}
+
+  liftIO $ do
+    T.IO.hPutStr outHandle $ addPrefixWhereNeeded content'
+    hFlush outHandle
+  where
+    addPrefixWhereNeeded :: T.Text -> T.Text
+    addPrefixWhereNeeded =
+      let prefix = makeJobMessagePrefix jobMessage
+          addPrefixAfterSubstr substr = T.intercalate (substr <> prefix) . T.splitOn substr
+       in addPrefixAfterSubstr "\r" . addPrefixAfterSubstr "\n"
```

Comment:
Why again are we splitting the `\r` and `\n` cases like this again, vs treating them as `\r\n`?

Reply:
Pfuf, those are good questions!
`\r` is valid because it can be used to "reset the cartridge" -> so it moves cursor to the start of the line removing previous output and writes what comes after it. It can be used for stuff like showing a progress bar, or percentages growing. Prisma uses it in some of its output, I think npm also possibly (e.g. when installing stuff).

As for windows -> I know they use \r\n in the files, but maybe they don't in the shell?
I would maybe just ignore this one for now, and we can handle that when the windows come :D.

Reply:
Ah, yes, silly Prisma-like executables. Forgot they do carriage return stuff and some other funky ANSI symbols. Ok, makes sense! Yes, let's punt on Windows for now! :D

Reply:
Hey `\r` is not that weird, we are also going to use it for `npm install` progress indicator probably :D!
Btw I am going to leave a TODO here to think more about Windows in the future.

Reply:
Haha, yeah it's true `\r` is fine for interactive outputs, I was just joking since I spaced on it. :D TODO for Windows makes sense. 👍🏻